core/blockchain.go

```diff
@@ -4139,12 +4163,196 @@ func (bc *BlockChain) SubscribeChain2HeadEvent(ch chan<- Chain2HeadEvent) event.
+	blocksSinceWindowStart := blockNum - bc.cacheWindowStart
+
+	if blocksSinceWindowStart >= bc.cacheWindowSize {
+		// Time to slide the window: discard active, promote next, create new next
+		bc.cacheLock.Lock()
+		oldActiveSize := len(bc.activeCacheMap)
+		bc.activeCacheMap = bc.nextCacheMap
+		bc.nextCacheMap = make(map[string]struct{})
+		bc.cacheWindowStart = blockNum
+		bc.cacheLock.Unlock()
+
+		log.Info("Sliding window: switched to next cache map",
+			"block", blockNum,
+			"newWindowStart", blockNum,
+			"discardedSize", oldActiveSize,
+			"newActiveSize", len(bc.activeCacheMap))
+		return true
+	}
+	return false
+}
+
+// GetCacheWindowStart returns the current cache window start block number.
+func (bc *BlockChain) GetCacheWindowStart() uint64 {
+	bc.cacheLock.RLock()
+	defer bc.cacheLock.RUnlock()
+	return bc.cacheWindowStart
+}
+
+// GetCacheWindowSize returns the cache window size.
+func (bc *BlockChain) GetCacheWindowSize() uint64 {
+	return bc.cacheWindowSize
+}
+
+// FilterWitnessWithSlidingCache filters a witness by removing state nodes present in the sliding window cache.
+// This is used when sending reduced witnesses to peers.
+func (bc *BlockChain) FilterWitnessWithSlidingCache(witness *stateless.Witness) (*stateless.Witness, int, int) {
+	if witness == nil {
+		return nil, 0, 0
+	}
+
+	filtered := witness.Copy()
+
+	bc.cacheLock.RLock()
+	defer bc.cacheLock.RUnlock()
+
+	originalStates := len(filtered.State)
+	filteredStates := make(map[string]struct{})
+	removed := 0
+
+	for stateNode := range filtered.State {
+		// Only include states NOT in active cache
+		if _, cached := bc.activeCacheMap[stateNode]; !cached {
+			filteredStates[stateNode] = struct{}{}
+		} else {
+			removed++
+		}
+	}
+
+	filtered.State = filteredStates
+
+	return filtered, originalStates, removed
+}
+
+// updateSlidingWindowCache updates the sliding window cache after processing a block.
+// This is used by both full nodes and witness-receiving nodes to maintain synchronized caches.
+func (bc *BlockChain) updateSlidingWindowCache(blockNum uint64, witness *stateless.Witness, statedb *state.StateDB) {
+	if witness == nil && statedb == nil {
+		return
+	}
+
+	bc.cacheLock.Lock()
+	defer bc.cacheLock.Unlock()
+
+	// Determine if we're in overlap period (should cache to both maps)
+	blocksSinceWindowStart := blockNum - bc.cacheWindowStart
+	inOverlapPeriod := blocksSinceWindowStart >= bc.cacheOverlapSize
+
+	cachedToActive := 0
+	cachedToNext := 0
+
+	// Cache witness state nodes if available
+	if witness != nil {
+		for stateNode := range witness.State {
+			// Add to active map
+			if _, exists := bc.activeCacheMap[stateNode]; !exists {
+				bc.activeCacheMap[stateNode] = struct{}{}
+				cachedToActive++
+			}
+			// Also add to next map if in overlap period
+			if inOverlapPeriod {
+				if _, exists := bc.nextCacheMap[stateNode]; !exists {
+					bc.nextCacheMap[stateNode] = struct{}{}
+					cachedToNext++
```

Comment:
It seems like the witness here is a super set of all trie nodes from active window start block to the current block (because of `mergeSpanCacheIntoWitness`) in stateless nodes. If this is true, `bc.nextCacheMap` will have the same size as `bc.activeCacheMap` and hence lose its original purpose - trim off unnecessary nodes. The correct implementation * I think * is to make sure the witness passed to `updateSlidingWindowCache` contains only the trie nodes that are required by the current block.

Reply:
Nice catch! Fixed [here](https://github.com/0xPolygon/bor/pull/1818/commits/61b1a40cedcb9e719aece0a5bdb1e82d86130c9a).